src/main/java/picard/analysis/directed/TargetMetricsCollector.java

```diff
@@ -548,63 +550,79 @@ public void acceptRecord(final SAMRecord record) {
                 rec = record;
             }
 
-            // Find the target overlaps
-            final Set<Interval> coveredTargets = new HashSet<>();
-            for (final AlignmentBlock block : rec.getAlignmentBlocks()) {
-                final int length = block.getLength();
-                final int refStart = block.getReferenceStart();
-                final int readStart = block.getReadStart();
-
-                for (int offset = 0; offset < length; ++offset) {
-                    final int refPos = refStart + offset;
-                    final int readPos = readStart + offset;
-                    final int qual = baseQualities[readPos - 1];
-
-                    if (qual <= 2) {
-                        metrics.PCT_EXC_BASEQ++;
-                        continue;
-                    }
+            // Calculate all the things that require examining individual bases in the read. This includes:
+            //   1. Per-base coverage
+            //   2. The number of reads contributing to per-base coverage per target
+            //   3. Unfiltered coverage information for het sensitivity
+            //   4. The count of bases rejected for being low baseq or off-target
+            //   5. The count of on-target bases and on-target bases from paired reads
+            final Set<Interval> coveredTargets = new HashSet<>(); // Each target is added to this the first time the read covers it
+            int readOffset = 0;
+            int refOffset  = rec.getAlignmentStart() -1;
+
+            for (final CigarElement cig : rec.getCigar()) {
+                final CigarOperator op = cig.getOperator();
+                final int len = cig.getLength();
+
+                for (int i=0; i<len; ++i) {
+                    if (op.isAlignment() || (this.includeIndels && op.isIndel())) {
```

Comment:
can this line be swapped with the previous line as it seems to encompass all of the for-loop and doesn't depend on `i`.

Reply:
The whole if is encompassed, but not the statements on lines 612 and 613 that update `readOffset` and `refOffset`.  I did at one point try inverting and adding an else that updated the offsets using the length of the element and saw no performance increase, so went back to how it is now.